### PR TITLE
[fix][txn] Port for master "Use PulsarClient HashWheelTimer to schedule producer batch trigger task"

### PR DESCRIPTION
**Pull request by lifepuzzlefun**
fix when create a lot of producer (> 10000) in the same jvm which will use eventLoop to do task check

Fixes https://github.com/apache/pulsar/issues/18055

this is a port for master version because 2.9.x is diverse from master

mr for 2.9.x https://github.com/apache/pulsar/pull/18058
- [x] `doc-not-needed`

### CODE REVIEW

**Review by coderzc — APPROVED** [review]

File: pulsar-client/src/main/java/org/apache/pulsar/client/impl/ProducerImpl.java [hunk 1]
```diff
@@ -159,7 +159,7 @@
     // The goal is to optimize batch density while also ensuring that a producer never waits longer than the configured
     // batchingMaxPublishDelayMicros to send a batch.
     // Only update from within synchronized block on this producer.
-    private ScheduledFuture<?> batchFlushTask;
+    private Timeout batchFlushTask;
```
Comment (tisonkun) [inline pulsar-client/src/main/java/org/apache/pulsar/client/impl/ProducerImpl.java:162]: The corresponding comment above should be updated also.

**Review by tisonkun — CHANGES_REQUESTED** [review]
Thanks for your contribution @lifepuzzlefun!

Comments inline.

File: pulsar-client/src/main/java/org/apache/pulsar/client/impl/ProducerImpl.java (hunk 1 above)
Comment (lifepuzzlefun) [reply pulsar-client/src/main/java/org/apache/pulsar/client/impl/ProducerImpl.java:162]: @tisonkun thanks for review, this change not change the logic for when trigger the schecule task. just change the scheduler  to hashwheeltimer for better performance. maybe the comment can remain the same.

Comment (tisonkun) [reply pulsar-client/src/main/java/org/apache/pulsar/client/impl/ProducerImpl.java:162]: Thanks for your explanation! SGTM.

**Review by tisonkun — APPROVED** [review]

**Comment by tisonkun** [comment]
```
  Error:  Tests run: 88, Failures: 1, Errors: 0, Skipped: 70, Time elapsed: 55.012 s <<< FAILURE! - in org.apache.pulsar.client.impl.TransactionEndToEndTest
  Error:  testSendTxnMessageTimeout(org.apache.pulsar.client.impl.TransactionEndToEndTest)  Time elapsed: 0.335 s  <<< FAILURE!
  java.lang.AssertionError: expected [true] but found [false]
  	at org.testng.Assert.fail(Assert.java:110)
  	at org.testng.Assert.failNotEquals(Assert.java:1413)
  	at org.testng.Assert.assertTrue(Assert.java:56)
  	at org.testng.Assert.assertTrue(Assert.java:66)
  	at org.apache.pulsar.client.impl.TransactionEndToEndTest.testSendTxnMessageTimeout(TransactionEndToEndTest.java:1368)
  	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
  	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
  	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
  	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
  	at org.testng.internal.invokers.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:139)
  	at org.testng.internal.invokers.InvokeMethodRunnable.runOne(InvokeMethodRunnable.java:47)
  	at org.testng.internal.invokers.InvokeMethodRunnable.call(InvokeMethodRunnable.java:76)
  	at org.testng.internal.invokers.InvokeMethodRunnable.call(InvokeMethodRunnable.java:11)
  	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
  	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
  	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
  	at java.base/java.lang.Thread.run(Thread.java:833)
```

It seems steady fail. Closing as stale for now...

I'm glad to nudge this patch if you can resolve the test failure and open a new PR.

**Pull request closed without merging**